File: android/core-imaging/src/main/java/com/offlinetracer/imaging/Denoise.kt
package com.offlinetracer.imaging

import kotlin.math.ceil
import kotlin.math.exp
import kotlin.math.max

/**
 * Noise reduction.
 *
 * Three filters are offered because they fail in different ways and the failure is what matters:
 * [bilateral] is right for sensor noise (it preserves the edges the tracer is looking for),
 * [median] is right for dust and salt-and-pepper on a scan (bilateral treats a dust speck as an
 * edge and protects it), and [anisotropicDiffusion] is right for suppressing woven or carved
 * texture that neither of the others touches.
 */
object Denoise {

    /**
     * Edge-preserving bilateral filter.
     *
     * Spatial radius is `ceil(2 * sigmaSpace)` — 2σ, not the 3σ used for a plain Gaussian, because
     * the range term already suppresses the tail and the kernel here is 2-D, so the outer ring costs
     * quadratically for a contribution the range weight has usually already zeroed.
     *
     * Non-positive sigmas return a copy.
     */
    fun bilateral(src: GrayF, sigmaSpace: Float, sigmaRange: Float): GrayF {
        if (sigmaSpace <= 0.05f || sigmaRange <= 0f) return src.copy()
        val w = src.width
        val h = src.height
        val s = src.data
        val r = max(1, ceil(2.0 * sigmaSpace).toInt())
        val side = 2 * r + 1

        val spatial = FloatArray(side * side)
        val invS = -1.0 / (2.0 * sigmaSpace * sigmaSpace)
        var si = 0
        for (dy in -r..r) {
            for (dx in -r..r) {
                spatial[si] = exp((dx * dx + dy * dy).toDouble() * invS).toFloat()
                si++
            }
        }
        // 256-entry range LUT indexed by round(|ΔI| * 255). Evaluating exp() per neighbour pair made
        // this single stage cost more than the rest of the pipeline put together on a 2048px image.
        val range = FloatArray(256)
        val invR = -1.0 / (2.0 * sigmaRange * sigmaRange)
        for (i in 0 until 256) {
            val d = i / 255.0
            range[i] = exp(d * d * invR).toFloat()
        }

        val out = FloatArray(w * h)
        // The documented hot spot, and the cleanest possible split: both lookup tables are finished and
        // read-only, `s` is never written, and each pixel's `(2r+1)²` accumulation lands in `out[i]`
        // alone. A share therefore evaluates the identical sum in the identical order for its own rows.
        Parallel.rows(h, Parallel.ROWS_NEIGHBOURHOOD) { fromY, toY ->
            for (y in fromY until toY) {
                val interiorY = y >= r && y + r < h
                val rowBase = y * w
                for (x in 0 until w) {
                    val i = rowBase + x
                    val c = s[i]
                    var wsum = 0f
                    var vsum = 0f
                    if (interiorY && x >= r && x + r < w) {
                        var ki = 0
                        var rowStart = i - r * w - r
                        for (dy in -r..r) {
                            var idx = rowStart
                            for (dx in -r..r) {
                                val v = s[idx]
                                var d = v - c
                                if (d < 0f) d = -d
                                var li = (d * 255f + 0.5f).toInt()
                                if (li > 255) li = 255
                                val ww = spatial[ki] * range[li]
                                wsum += ww
                                vsum += ww * v
                                ki++
                                idx++
                            }
                            rowStart += w
                        }
                    } else {
                        var ki = 0
                        for (dy in -r..r) {
                            var yy = y + dy
                            if (yy < 0) yy = 0
                            if (yy > h - 1) yy = h - 1
                            val base = yy * w
                            for (dx in -r..r) {
                                var xx = x + dx
                                if (xx < 0) xx = 0
                                if (xx > w - 1) xx = w - 1
                                val v = s[base + xx]
                                var d = v - c
                                if (d < 0f) d = -d
                                var li = (d * 255f + 0.5f).toInt()
                                if (li > 255) li = 255
                                val ww = spatial[ki] * range[li]
                                wsum += ww
                                vsum += ww * v
                                ki++
                            }
                        }
                    }
                    out[i] = if (wsum > 0f) vsum / wsum else c
                }
            }
        }
        return GrayF(w, h, out)
    }

    /**
     * Median filter over a `(2r+1)²` window, edge-clamped. A non-positive radius returns a copy.
     *
     * Radius 1 sorts the nine exact float samples. Radius ≥ 2 slides a 256-bin histogram across each
     * row, which makes the cost independent of the window area — the input at this point in the
     * pipeline came from 8-bit pixels, so the histogram is lossless for real sources and the
     * alternative (sorting 25+ floats per pixel) is minutes rather than seconds on a large scan.
     */
    fun median(src: GrayF, radius: Int): GrayF {
        if (radius <= 0) return src.copy()
        return if (radius == 1) median3(src) else medianHistogram(src, radius)
    }

    /**
     * Perona–Malik anisotropic diffusion with conduction `c(g) = exp(-(g/κ)²)`.
     *
     * [lambda] is the explicit-scheme step; 0.25 is the stability limit for a 4-neighbourhood and
     * larger values oscillate, so it is clamped there. Non-positive [iterations] or [kappa] return
     * a copy.
     */
    fun anisotropicDiffusion(src: GrayF, iterations: Int, kappa: Float, lambda: Float = 0.25f): GrayF {
        if (iterations <= 0 || kappa <= 0f) return src.copy()
        val w = src.width
        val h = src.height
        val lam = Px.clamp(lambda, 0f, 0.25f)
        val invK2 = -1.0 / (kappa.toDouble() * kappa)
        var cur = src.data.copyOf()
        var next = FloatArray(w * h)
        for (pass in 0 until iterations) {
            // Rows within one diffusion step are split; the step loop itself is not, and the double
            // buffer is what makes that legal — every pixel reads `rd` (finished last step) and writes
            // only `wr[i]`. Diffusing in place would make each pixel see whichever of its neighbours
            // its own share had already updated, i.e. an answer that depends on the band boundaries.
            //
            // Bound to locals rather than captured because `cur`/`next` are `var`s and a captured `var`
            // becomes a heap `Ref` read four times per pixel.
            val rd = cur
            val wr = next
            Parallel.rows(h, Parallel.ROWS_NEIGHBOURHOOD) { fromY, toY ->
                for (y in fromY until toY) {
                    val base = y * w
                    val up = if (y > 0) base - w else base
                    val dn = if (y < h - 1) base + w else base
                    for (x in 0 until w) {
                        val i = base + x
                        val v = rd[i]
                        val gN = rd[up + x] - v
                        val gS = rd[dn + x] - v
                        val gW = rd[base + (if (x > 0) x - 1 else 0)] - v
                        val gE = rd[base + (if (x < w - 1) x + 1 else w - 1)] - v
                        val cN = exp(gN.toDouble() * gN * invK2)
                        val cS = exp(gS.toDouble() * gS * invK2)
                        val cW = exp(gW.toDouble() * gW * invK2)
                        val cE = exp(gE.toDouble() * gE * invK2)
                        wr[i] = v + lam * (cN * gN + cS * gS + cW * gW + cE * gE).toFloat()
                    }
                }
            }
            val swap = cur
            cur = next
            next = swap
        }
        return GrayF(w, h, cur)
    }

    /**
     * Replace a pixel by the local median only where it differs from that median by more than
     * [threshold]; everything else is passed through untouched.
     *
     * This is what a dust-removal pass should do. A plain median softens every stroke in the image
     * to remove a handful of specks, and on faint pencil work that loss is not recoverable.
     */
    fun despeckle(src: GrayF, radius: Int, threshold: Float): GrayF {
        if (radius <= 0) return src.copy()
        val med = median(src, radius)
        val s = src.data
        val m = med.data
        val out = FloatArray(s.size)
        for (i in s.indices) {
            var d = s[i] - m[i]
            if (d < 0f) d = -d
            out[i] = if (d > threshold) m[i] else s[i]
        }
        return GrayF(src.width, src.height, out)
    }

    private fun median3(src: GrayF): GrayF {
        val w = src.width
        val h = src.height
        val out = FloatArray(w * h)
        Parallel.rows(h, Parallel.ROWS_KERNEL) { fromY, toY ->
            // The sort scratch is allocated **per share**. Hoisting it out of the lambda the way the
            // single-threaded version did would have every thread sorting into the same nine floats,
            // and the corruption would look like plausible speckle rather than like a bug.
            val buf = FloatArray(9)
            for (y in fromY until toY) {
                for (x in 0 until w) {
                    var c = 0
                    for (dy in -1..1) {
                        for (dx in -1..1) {
                            buf[c] = src.clamped(x + dx, y + dy)
                            c++
                        }
                    }
                    for (i in 1 until 9) {
                        val v = buf[i]
                        var j = i - 1
                        while (j >= 0 && buf[j] > v) {
                            buf[j + 1] = buf[j]
                            j--
                        }
                        buf[j + 1] = v
                    }
                    out[y * w + x] = buf[4]
                }
            }
        }
        return GrayF(w, h, out)
    }

    private fun medianHistogram(src: GrayF, r: Int): GrayF {
        val w = src.width
        val h = src.height
        val n = w * h
        val q = IntArray(n)
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            for (i in from until to) q[i] = Px.toByte255(src.data[i])
        }
        val out = FloatArray(n)
        val side = 2 * r + 1
        val windowSize = side * side
        val rank = windowSize / 2
        // The sliding window is a *within-row* dependency, not a cross-row one: the histogram is rebuilt
        // from scratch at the start of every row, so a row's answer is a function of `q` alone and rows
        // split cleanly. Each share needs its own 256 bins for the same reason [median3] needs its own
        // nine floats.
        Parallel.rows(h, Parallel.ROWS_KERNEL) { fromY, toY ->
            val hist = IntArray(256)
            for (y in fromY until toY) {
                java.util.Arrays.fill(hist, 0)
                for (dy in -r..r) {
                    var yy = y + dy
                    if (yy < 0) yy = 0
                    if (yy > h - 1) yy = h - 1
                    val base = yy * w
                    for (dx in -r..r) {
                        var xx = dx
                        if (xx < 0) xx = 0
                        if (xx > w - 1) xx = w - 1
                        hist[q[base + xx]]++
                    }
                }
                // Huang's moving median: track the bin holding the median and the count strictly below
                // it, so a step costs O(window width) instead of a 256-bin rescan per pixel.
                var mdn = 0
                var below = 0
                while (below + hist[mdn] <= rank) {
                    below += hist[mdn]
                    mdn++
                }
                out[y * w] = mdn * (1f / 255f)
                for (x in 1 until w) {
                    var outCol = x - r - 1
                    if (outCol < 0) outCol = 0
                    if (outCol > w - 1) outCol = w - 1
                    var inCol = x + r
                    if (inCol < 0) inCol = 0
                    if (inCol > w - 1) inCol = w - 1
                    for (dy in -r..r) {
                        var yy = y + dy
                        if (yy < 0) yy = 0
                        if (yy > h - 1) yy = h - 1
                        val base = yy * w
                        val ov = q[base + outCol]
                        hist[ov]--
                        if (ov < mdn) below--
                        val iv = q[base + inCol]
                        hist[iv]++
                        if (iv < mdn) below++
                    }
                    while (below > rank) {
                        mdn--
                        below -= hist[mdn]
                    }
                    while (below + hist[mdn] <= rank) {
                        below += hist[mdn]
                        mdn++
                    }
                    out[y * w + x] = mdn * (1f / 255f)
                }
            }
        }
        return GrayF(w, h, out)
    }
}

File: frontend/lib/trace/engine/denoise.ts
import { GrayF, Px } from './buffers';

/**
 * Denoising. See ALGORITHMS.md §4.
 *
 * Bilateral and median are both offered because they fail differently: bilateral is the right tool
 * for sensor noise, median for salt-and-pepper dust on a scan. Picking one for the user would make
 * the other class of source untraceable.
 */

/**
 * Edge-preserving bilateral filter, edge-clamped.
 *
 * The range term is read from a 256-entry LUT indexed by `round(|dI| * 255)`; recomputing `exp` per
 * neighbour pair made this stage dominate the whole pipeline. Spatial radius is `ceil(2 * sigmaSpace)`
 * — 2σ not 3σ, because the range term already suppresses the tail and the kernel is 2-D.
 *
 * @returns a new GrayF; `src.copy()` when either sigma is <= 0.
 */
export function bilateral(src: GrayF, sigmaSpace: number, sigmaRange: number): GrayF {
  if (sigmaSpace <= 0 || sigmaRange <= 0) return src.copy();
  const w = src.width;
  const h = src.height;
  const d = src.data;
  const r = Math.max(1, Math.ceil(2 * sigmaSpace));
  const side = 2 * r + 1;

  const spatial = new Float32Array(side * side);
  const invSpace = 1 / (2 * sigmaSpace * sigmaSpace);
  for (let dy = -r; dy <= r; dy++) {
    for (let dx = -r; dx <= r; dx++) {
      spatial[(dy + r) * side + (dx + r)] = Math.exp(-(dx * dx + dy * dy) * invSpace);
    }
  }
  const rangeLut = new Float32Array(256);
  const invRange = 1 / (2 * sigmaRange * sigmaRange);
  for (let i = 0; i < 256; i++) {
    const t = i / 255;
    rangeLut[i] = Math.exp(-(t * t) * invRange);
  }

  const out = new Float32Array(w * h);
  for (let y = 0; y < h; y++) {
    const row = y * w;
    for (let x = 0; x < w; x++) {
      const centre = d[row + x];
      let acc = 0;
      let wsum = 0;
      for (let dy = -r; dy <= r; dy++) {
        const sy = y + dy;
        const srow = (sy < 0 ? 0 : sy >= h ? h - 1 : sy) * w;
        const krow = (dy + r) * side;
        for (let dx = -r; dx <= r; dx++) {
          const sx = x + dx;
          const v = d[srow + (sx < 0 ? 0 : sx >= w ? w - 1 : sx)];
          let idx = Math.round(Math.abs(v - centre) * 255);
          if (idx > 255) idx = 255;
          const weight = spatial[krow + dx + r] * rangeLut[idx];
          acc += weight * v;
          wsum += weight;
        }
      }
      out[row + x] = wsum > 0 ? acc / wsum : centre;
    }
  }
  return new GrayF(w, h, out);
}

/**
 * Median filter, edge-clamped.
 *
 * `radius == 1` sorts the 9 samples exactly. `radius >= 2` uses a sliding 256-bin histogram, which is
 * O(1) amortised in the window width; the price is that the output is quantised to 1/255 steps. That
 * is the trade the reference algorithm makes and both engines make it identically, because a median
 * that costs O(r^2) makes radius 6 unusable on a phone.
 *
 * @returns a new GrayF; `src.copy()` when `radius <= 0`.
 */
export function median(src: GrayF, radius: number): GrayF {
  const r = radius | 0;
  if (r <= 0) return src.copy();
  return r === 1 ? median3(src) : medianHistogram(src, r);
}

function median3(src: GrayF): GrayF {
  const w = src.width;
  const h = src.height;
  const d = src.data;
  const out = new Float32Array(w * h);
  const buf = new Float32Array(9);
  for (let y = 0; y < h; y++) {
    const ym = (y > 0 ? y - 1 : 0) * w;
    const y0 = y * w;
    const yp = (y < h - 1 ? y + 1 : h - 1) * w;
    for (let x = 0; x < w; x++) {
      const xm = x > 0 ? x - 1 : 0;
      const xp = x < w - 1 ? x + 1 : w - 1;
      buf[0] = d[ym + xm];
      buf[1] = d[ym + x];
      buf[2] = d[ym + xp];
      buf[3] = d[y0 + xm];
      buf[4] = d[y0 + x];
      buf[5] = d[y0 + xp];
      buf[6] = d[yp + xm];
      buf[7] = d[yp + x];
      buf[8] = d[yp + xp];
      // Insertion sort: nine elements, no allocation, and it beats every clever network in practice.
      for (let i = 1; i < 9; i++) {
        const v = buf[i];
        let j = i - 1;
        while (j >= 0 && buf[j] > v) {
          buf[j + 1] = buf[j];
          j--;
        }
        buf[j + 1] = v;
      }
      out[y0 + x] = buf[4];
    }
  }
  return new GrayF(w, h, out);
}

function medianHistogram(src: GrayF, r: number): GrayF {
  const w = src.width;
  const h = src.height;
  const out = new Float32Array(w * h);
  const bins = new Int32Array(256);
  const q = new Uint8Array(w * h);
  const d = src.data;
  for (let i = 0; i < d.length; i++) q[i] = Px.toByte255(d[i]);
  const win = (2 * r + 1) * (2 * r + 1);
  const half = (win + 1) >> 1;

  for (let y = 0; y < h; y++) {
    bins.fill(0);
    // Prime the window at x = 0, columns [-r, r] with out-of-range columns clamped to column 0.
    for (let dy = -r; dy <= r; dy++) {
      const sy = y + dy;
      const srow = (sy < 0 ? 0 : sy >= h ? h - 1 : sy) * w;
      for (let dx = -r; dx <= r; dx++) {
        const sx = dx < 0 ? 0 : dx >= w ? w - 1 : dx;
        bins[q[srow + sx]]++;
      }
    }
    out[y * w] = pickMedian(bins, half) / 255;
    for (let x = 1; x < w; x++) {
      const addX = x + r;
      const subX = x - r - 1;
      const ax = addX >= w ? w - 1 : addX;
      const sx = subX < 0 ? 0 : subX;
      for (let dy = -r; dy <= r; dy++) {
        const sy = y + dy;
        const srow = (sy < 0 ? 0 : sy >= h ? h - 1 : sy) * w;
        bins[q[srow + ax]]++;
        bins[q[srow + sx]]--;
      }
      out[y * w + x] = pickMedian(bins, half) / 255;
    }
  }
  return new GrayF(w, h, out);
}

function pickMedian(bins: Int32Array, half: number): number {
  let cum = 0;
  for (let i = 0; i < 256; i++) {
    cum += bins[i];
    if (cum >= half) return i;
  }
  return 255;
}

/**
 * Perona-Malik anisotropic diffusion with conduction function `c(g) = exp(-(g/kappa)^2)`.
 *
 * @param iterations fixed count; 0 or fewer returns `src.copy()`
 * @param kappa      conduction threshold; <= 0 returns `src.copy()`
 * @param lambda     step size, clamped to 0.25 which is the stability limit for a 4-neighbourhood
 * @returns a new GrayF the same size as `src`.
 */
export function anisotropicDiffusion(
  src: GrayF,
  iterations: number,
  kappa: number,
  lambda = 0.25,
): GrayF {
  const iters = iterations | 0;
  if (iters <= 0 || kappa <= 0) return src.copy();
  const lam = Px.clamp(lambda, 0, 0.25);
  const w = src.width;
  const h = src.height;
  let cur = src.data.slice();
  let next = new Float32Array(w * h);
  const invK2 = 1 / (kappa * kappa);
  for (let it = 0; it < iters; it++) {
    for (let y = 0; y < h; y++) {
      const row = y * w;
      const rowN = (y > 0 ? y - 1 : 0) * w;
      const rowS = (y < h - 1 ? y + 1 : h - 1) * w;
      for (let x = 0; x < w; x++) {
        const c = cur[row + x];
        const dn = cur[rowN + x] - c;
        const ds = cur[rowS + x] - c;
        const de = cur[row + (x < w - 1 ? x + 1 : w - 1)] - c;
        const dw = cur[row + (x > 0 ? x - 1 : 0)] - c;
        next[row + x] =
          c +
          lam *
            (Math.exp(-dn * dn * invK2) * dn +
              Math.exp(-ds * ds * invK2) * ds +
              Math.exp(-de * de * invK2) * de +
              Math.exp(-dw * dw * invK2) * dw);
      }
    }
    const swap = cur;
    cur = next;
    next = swap;
  }
  return new GrayF(w, h, cur);
}

/**
 * Median only where the pixel disagrees with its neighbourhood by more than `threshold`.
 *
 * This is the correct shape for dust removal: a plain median softens every stroke it passes over,
 * while despeckle leaves the 99% of pixels that agree with their surroundings bit-identical.
 *
 * @returns a new GrayF; `src.copy()` when `radius <= 0`.
 */
export function despeckle(src: GrayF, radius: number, threshold: number): GrayF {
  if (radius <= 0) return src.copy();
  const med = median(src, radius);
  const n = src.data.length;
  const out = new Float32Array(n);
  const a = src.data;
  const m = med.data;
  const t = Math.abs(threshold);
  for (let i = 0; i < n; i++) {
    const v = a[i];
    out[i] = Math.abs(v - m[i]) > t ? m[i] : v;
  }
  return new GrayF(src.width, src.height, out);
}
